src/DynamoPackages/PackageUploadBuilder.cs

```diff
@@ -94,15 +94,15 @@ public PackageUpload NewPackageUpload(Package package, string packagesDirectory,
         /// <param name="handle"></param>
         /// <returns></returns>
         /// <exception cref="ArgumentNullException"></exception>
-        public PackageUpload NewPackageRetainUpload(Package package, string packagesDirectory, IEnumerable<IEnumerable<string>> files, IEnumerable<string> markdownFiles, PackageUploadHandle handle)
+        public PackageUpload NewPackageRetainUpload(Package package, string packagesDirectory, IEnumerable<string> roots, IEnumerable<IEnumerable<string>> files, IEnumerable<string> markdownFiles, PackageUploadHandle handle)
```

Comment:
@QilongTang is it OK to break public API here? I know "retain" is a newer API and may not be leveraged by many users currently. Just wanted to start this discussion.

We can either override or make a note in release notes.

Reply:
We probably want to override with a new version and use that one instead, thank you for pointing it out

Reply:
Oh, of course! I did not think about the API breaking change here, thank you for pointing it out @zeusongit . I will try to sort that out.

Reply:
I recreated the obsolete logic chain to prevent breaking API, marking it [Obsolete]. I don't think I had another option but to have the whole of the old methods in, as the signature was being passed onwards continuously.